### ISSUE TITLE
Upgrading from 2.7 to 3.1 fails

### ISSUE THREAD

**Issue by emmericp** [ceph/ceph-iscsi#117]
upgrading from 2.7 to 3.1 leaves me with a gateway.conf with no gateway entries:

this problem is new with config v10, our builds on 4fb1f0ccb3d3fd039443a9d6f16f458946633598 work just fine

```
{
    "created": "2019/06/28 15:13:19",
    "discovery_auth": {
        "mutual_password": "",
        "mutual_password_encryption_enabled": false,
        "mutual_username": "",
        "password": "",
        "password_encryption_enabled": false,
        "username": ""
    },
    "disks": {
        "rbd/test": {
            "backstore": "user:rbd",
            "backstore_object_name": "rbd.test",
            "controls": {},
            "created": "2019/06/28 15:14:36",
            "image": "test",
            "owner": "server-2",
            "pool": "rbd",
            "pool_id": 1,
            "updated": "2019/06/28 21:27:18",
            "wwn": "ca848c29-00b3-423a-ac79-3273e5b62075"
        },
        "rbd/test2": {
            "backstore": "user:rbd",
            "backstore_object_name": "rbd.test2",
            "controls": {},
            "created": "2019/06/28 15:14:42",
            "image": "test2",
            "owner": "server-1",
            "pool": "rbd",
            "pool_id": 1,
            "updated": "2019/06/28 21:27:18",
            "wwn": "8bf6c1f0-2fba-497f-8281-1d6b1b933711"
        }
    },
    "epoch": 1381,
    "gateways": {},
    "targets": {
        "iqn.2017-01.io.croit.iscsi:ceph-gateway": {
            "acl_enabled": true,
            "clients": {
                "iqn.1998-01.com.vmware:esxi-0e1df5fb": {
                    "auth": {
                        "mutual_password": "",
                        "mutual_password_encryption_enabled": false,
                        "mutual_username": "",
                        "password": "testtesttest",
                        "password_encryption_enabled": false,
                        "username": "testtest"
                    },
                    "group_name": "",
                    "luns": {
                        "rbd/test": {
                            "lun_id": 0
                        }
                    }
                },
                "iqn.2111-02.0.0.2.4:lol2": {
                    "auth": {
                        "mutual_password": "",
                        "mutual_password_encryption_enabled": false,
                        "mutual_username": "",
                        "password": "testtesttest",
                        "password_encryption_enabled": false,
                        "username": "testtest"
                    },
                    "group_name": "",
                    "luns": {
                        "rbd/test": {
                            "lun_id": 0
                        }
                    }
                }
            },
            "controls": {},
            "created": "2019/06/28 21:03:52",
            "disks": [
                "rbd/test2",
                "rbd/test"
            ],
            "groups": {},
            "ip_list": [
                "10.0.0.101",
                "10.0.0.102"
            ],
            "portals": {},
            "updated": "2019/06/28 21:27:18"
        }
    },
    "updated": "2019/06/28 21:27:18",
    "version": 10
}
```

**Comment by dillaman** [comment]
Any chance you have your pre-upgrade config available?

**Comment by mikechristie** [comment]
It looks like the gw fqdn updater could not figure out the domain name so it did not update the gateways. The lun owner value is stuck as the host name only.


Is your networking/dns/hostname setup so that the nodes can have a fully qualified domain name or do you by any chance just have a local only setup and instead of doing something like myhost.something.local you just have myhost?

**Comment by mikechristie** [comment]
Or do you also maybe just have myhost.local?

**Comment by emmericp** [comment]
Running a local test , i.e. hostname -s and -f return the same short hostname.

Nothing special in the old config, but I didn‘t save it

**Comment by mikechristie** [comment]
It’s just a local setup with no domain name setup, correct? Or is hostname -f incorrectly retuning only the host name? The Ceph iscsi tools/daemon perform the same syscalls so that is why I am asking if the system is just not setup with a domain or if hostname -f is reporting the wrong info.

@ricardoasmarques,

It looks like we have to have a fqdn (it can just be a local one like bogus.local) or the code will fail. This was intentional right?

**Comment by emmericp** [comment]
```
$ python
>>> import socket
>>> print(socket.getfqdn())
server-1
```

/etc/hosts:

```
127.0.0.1 localhost
10.0.0.1 server-1
```

**Comment by dillaman** [comment]
I opened PR #118 to hopefully prevent this issue in the future. There is a WIP build that includes this fix located here [1] if you are able to test it.

[1] https://shaman.ceph.com/repos/ceph-iscsi/wip-fqdn-upgrade-fix/826f788dcacf27a437f64b0d79d1394bacf8a68f/default/131333/

**Comment by mikechristie** [comment]
Jason, I replicated the bug here, tested your patch, merged it and made a new release 3.2 hoping other people will see 3.2 and not get their setups hosed.

@emmericp  

To start up your system again you are going to to have to do the following:

- Stop the rbd-target-api daemons on all gateways.

# systemctl stop rbd-target-api

- Install ceph-iscsi-3.2 on all the gateways.

- Delete the gateway.conf. From one of the gateways do:

# rados -p rbd rm gateway.conf

- Restart all the systems/daemon with ceph-iscsi-3.2.

Either reboot the system with your normal reboot command or start the daemon

systemctl start tbd-target-api

- Re-add gateways, targets, images, etc with gwcli.

For the images since they are already created, you just need to add them by not passing in the size argument. Instead of doing:

/disks> create rbd test1 100G

do

/disks> create rbd test1

**Comment by emmericp** [comment]
thanks for the quick fix!
